ci(docs-infra): run with Ivy

gkalpak opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~


## PR Type
- [x] angular.io application / infrastructure changes


## What is the new behavior?
Update `aio` to run with Ivy and enable more `aio` tests to be run on CI with Ivy.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
Jira issue: [AVM-15](https://angular-team.atlassian.net/browse/AVM-15)

Blocked on:
- [x] ~~#28773~~
- [x] ~~#28963~~
- [x] ~~#29209~~
- [x] ~~#29602~~
- [x] ~~#29989~~
- [ ] ~~#29978~~ (superceded by #30022)
- [ ] ~~#30022~~ (superceded by angular/components#15898)
- [x] ~~angular/components#15898~~
- [x] ~~#30037~~
- [x] ~~#30121~~ (fixed in this PR)

mary-poppins commented:
You can preview 70de12b at https://pr28530-70de12b.ngbuilds.io/.
You can preview 17037e3 at https://pr28530-17037e3.ngbuilds.io/.

mary-poppins commented:
You can preview d937ad4 at https://pr28530-d937ad4.ngbuilds.io/.

mary-poppins commented:
You can preview a0829ad at https://pr28530-a0829ad.ngbuilds.io/.

mary-poppins commented:
You can preview e913f1d at https://pr28530-e913f1d.ngbuilds.io/.

mary-poppins commented:
You can preview c856125 at https://pr28530-c856125.ngbuilds.io/.

mary-poppins commented:
You can preview c3e985d at https://pr28530-c3e985d.ngbuilds.io/.

mary-poppins commented:
You can preview dc4ffd8 at https://pr28530-dc4ffd8.ngbuilds.io/.

mary-poppins commented:
You can preview 28a5afd at https://pr28530-28a5afd.ngbuilds.io/.

mary-poppins commented:
You can preview 0920a4b at https://pr28530-0920a4b.ngbuilds.io/.

mary-poppins commented:
You can preview 68b49d4 at https://pr28530-68b49d4.ngbuilds.io/.

mary-poppins commented:
You can preview dbca601 at https://pr28530-dbca601.ngbuilds.io/.

mary-poppins commented:
You can preview 1a1966c at https://pr28530-1a1966c.ngbuilds.io/.

mary-poppins commented:
You can preview a063924 at https://pr28530-a063924.ngbuilds.io/.

mary-poppins commented:
You can preview 2c28e37 at https://pr28530-2c28e37.ngbuilds.io/.

filipesilva commented on aio/scripts/switch-to-ivy.js:
```diff
@@ -10,37 +10,36 @@ set('-e');
+  // Run ngcc.
+  const ngccArgs = '--loglevel debug --properties es2015 module';
+  console.log(`\nRunning ngcc (with args: ${ngccArgs})...`);
+  exec(`yarn ivy-ngcc ${ngccArgs}`);
```
I think the latest CLI will do all ngcc'ing automatically (https://github.com/angular/angular-cli/pull/13968). I checked with @alan-agius4 and it will happen as long as the Ivy flag is in the tsconfig, nothing else is needed. Can you try it?

gkalpak replied:
Once #29926 is merged, I will give it a go.

mary-poppins commented:
You can preview b833b2f at https://pr28530-b833b2f.ngbuilds.io/.

mary-poppins commented:
You can preview 8faa0f0 at https://pr28530-8faa0f0.ngbuilds.io/.

mary-poppins commented:
You can preview e1ccd4c at https://pr28530-e1ccd4c.ngbuilds.io/.

mary-poppins commented:
You can preview 779b670 at https://pr28530-779b670.ngbuilds.io/.

mary-poppins commented:
You can preview ec81d63 at https://pr28530-ec81d63.ngbuilds.io/.

mary-poppins commented:
You can preview 0cd69ae at https://pr28530-0cd69ae.ngbuilds.io/.

mary-poppins commented:
You can preview 6f6d37b at https://pr28530-6f6d37b.ngbuilds.io/.

mary-poppins commented:
You can preview f26d9a9 at https://pr28530-f26d9a9.ngbuilds.io/.

mary-poppins commented:
You can preview 2852057 at https://pr28530-2852057.ngbuilds.io/.

mary-poppins commented:
You can preview 3788918 at https://pr28530-3788918.ngbuilds.io/.

mary-poppins commented:
You can preview e953b9b at https://pr28530-e953b9b.ngbuilds.io/.

gkalpak committed:
fix(docs-infra): update app code to work with Ivy

This commit also enables more tests to be run on CI with Ivy.

gkalpak committed:
fix(ivy): allow `R3TestBedCompiler` to work in ngcc-processed apps

Previously, `R3TestBedCompiler` was dynamically defining an
`@NgModule`-decorated `CompilerModule` class inside a method call.
Since ngcc only processes top-level classes, this class was not
transformed causing failures in unit tests (see #30121 for details).

This commit fixes it by using `compileNgModuleDefs()` directly (similar
to the fix in #30037).

Fixes #30121

mary-poppins commented:
You can preview ef63e75 at https://pr28530-ef63e75.ngbuilds.io/.

mary-poppins commented:
You can preview 3105d83 at https://pr28530-3105d83.ngbuilds.io/.

mary-poppins commented:
You can preview 715db19 at https://pr28530-715db19.ngbuilds.io/.

mary-poppins commented:
You can preview aa760f3 at https://pr28530-aa760f3.ngbuilds.io/.

mary-poppins commented:
You can preview 1f9b0f0 at https://pr28530-1f9b0f0.ngbuilds.io/.

mary-poppins commented:
You can preview f7a127c at https://pr28530-f7a127c.ngbuilds.io/.

gkalpak committed:
fixup! fix(docs-infra): update app code to work with Ivy

mary-poppins commented:
You can preview 187b7b4 at https://pr28530-187b7b4.ngbuilds.io/.

gkalpak committed:
fixup! fix(docs-infra): update app code to work with Ivy

mary-poppins commented:
You can preview 2fe38cc at https://pr28530-2fe38cc.ngbuilds.io/.

gkalpak commented:
**merge-assistance:** Requires g3sync presubmit.
**merge-assistance:** This PR doesn't require all the requested reviews any more (many things have been split out to separate PRs).

kara reviewed (approved):
LGTM (global)

kara commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:246212768:BASE:246213135:1556750384468:d9b5e860)

IgorMinar commented on aio/src/app/custom-elements/code/code-tabs.component.spec.ts:
```diff
@@ -23,10 +23,10 @@ describe('CodeTabsComponent', () => {
     fixture = TestBed.createComponent(HostComponent);
+    fixture.detectChanges();
+
     hostComponent = fixture.componentInstance;
```
It's this a breaking change?

gkalpak replied:
The `componentInstance` is still available before `detectChanges()`. It is the `codeTabsComponent` (which is a `ViewChild` query) that isn't.

This is related to the (known breaking) changes in `ViewChild` queries. We could have made [the `codeTabsComponent` `ViewChild`][1] static instead (and leave the test as is).

[1]: https://github.com/gkalpak/angular/blob/2fe38cc445ded46e48eda8baac585e20f297c988/aio/src/app/custom-elements/code/code-tabs.component.spec.ts#L95

IgorMinar reviewed (commented):
Doesn't updating the app code imply a breaking change?

gkalpak commented:
> Doesn't updating the app code imply a breaking change?

I assume you are referring to the [changes in `SearchBoxComponent`][1]. These are related to the (known breaking) changes in `ViewChild` queries.

I had originally left [the `searchBox` query][2] as non-static in this PR and moved the logic from `ngOnInit()` to `ngAfterViewInit()`. The query was changed to static in another PR, so the change from `ngOnInit()` to `ngAfterViewInit()` was not any more necessary (but I missed it when rebasing the PR).

Anyway, I am going to leave it as is, since it doesn't make any difference (and I generally prefer accessing view query results in `ngAfterViewInit()`), but the good news is that there is no breaking change we didn't already know :grin:

[1]: https://github.com/angular/angular/pull/28530/files?file-filters%5B%5D=.ts#diff-31dc5ce2059811d5b6b4a25b6705a9d5
[2]: https://github.com/gkalpak/angular/blob/2fe38cc445ded46e48eda8baac585e20f297c988/aio/src/app/search/search-box/search-box.component.ts#L32
